storage/statedb/hasher.go

```diff
@@ -148,6 +187,58 @@ func (h *hasher) hashChildren(original node, db *Database) (node, node) {
 	}
 }
 
+type hashResult struct {
+	index     int
+	collapsed node
+	cached    node
+}
+
+func (h *hasher) hashChildrenFromRoot(original node, db *Database) (node, node) {
```

Comment:
`hashChildrenFromRoot` and `hashChildren` are 90% identical. It would be nice to implement them as a function. The same is true for `hashRoot` and `hash`.

Reply:
If we don't need to maintain the previous serial execution code, `hashChildren` will be deleted.